File: frontend/src/utils/errorCodes.ts
/**
 * Machine-readable error codes matching the backend ErrorCode constants.
 *
 * These codes are returned in the `errors[].code` field of the API envelope.
 * The `getErrorMessage()` helper maps them to user-friendly i18n keys.
 */

// Authentication
export const AUTH_INVALID_CREDENTIALS = "AUTH_INVALID_CREDENTIALS";
export const AUTH_TOKEN_EXPIRED = "AUTH_TOKEN_EXPIRED";
export const AUTH_TOKEN_INVALID = "AUTH_TOKEN_INVALID";
export const AUTH_ACCOUNT_INACTIVE = "AUTH_ACCOUNT_INACTIVE";
export const AUTH_EMAIL_TAKEN = "AUTH_EMAIL_TAKEN";
export const AUTH_PASSWORD_EXPIRED = "AUTH_PASSWORD_EXPIRED";
export const AUTH_PASSWORD_MISMATCH = "AUTH_PASSWORD_MISMATCH";
export const AUTH_PASSWORD_UNAVAILABLE = "AUTH_PASSWORD_UNAVAILABLE";
export const AUTH_OAUTH_NOT_CONFIGURED = "AUTH_OAUTH_NOT_CONFIGURED";
export const AUTH_OAUTH_NOT_IMPLEMENTED = "AUTH_OAUTH_NOT_IMPLEMENTED";
export const AUTH_MISSING_AUTH_CODE = "AUTH_MISSING_AUTH_CODE";

// Authorization
export const AUTHZ_ADMIN_REQUIRED = "AUTHZ_ADMIN_REQUIRED";
export const AUTHZ_FORBIDDEN = "AUTHZ_FORBIDDEN";
export const AUTHZ_API_KEY_INVALID = "AUTHZ_API_KEY_INVALID";

// Credits
export const CREDIT_INSUFFICIENT = "CREDIT_INSUFFICIENT";

// Resources
export const RESOURCE_NOT_FOUND = "RESOURCE_NOT_FOUND";
export const RESOURCE_CONFLICT = "RESOURCE_CONFLICT";
export const RESOURCE_GONE = "RESOURCE_GONE";

// Validation
export const VALIDATION_ERROR = "VALIDATION_ERROR";
export const VALIDATION_EMPTY_UPDATE = "VALIDATION_EMPTY_UPDATE";
export const VALIDATION_LIMIT_EXCEEDED = "VALIDATION_LIMIT_EXCEEDED";

// External services
export const EXTERNAL_TIMEOUT = "EXTERNAL_TIMEOUT";
export const EXTERNAL_FAILURE = "EXTERNAL_FAILURE";
export const EXTERNAL_PROVIDER_UNAVAILABLE = "EXTERNAL_PROVIDER_UNAVAILABLE";

// Server
export const INTERNAL_ERROR = "INTERNAL_ERROR";

// Client-side (not from backend)
export const NETWORK_ERROR = "NETWORK_ERROR";
export const TIMEOUT = "TIMEOUT";

/**
 * Map from error code to i18n key under "errorCodes.*".
 *
 * If a code is not in the map, `getErrorMessage` falls back to the raw
 * server message or a generic "Unknown error" string.
 */
const ERROR_CODE_I18N_MAP: Record<string, string> = {
  [AUTH_INVALID_CREDENTIALS]: "errorCodes.authInvalidCredentials",
  [AUTH_TOKEN_EXPIRED]: "errorCodes.authTokenExpired",
  [AUTH_TOKEN_INVALID]: "errorCodes.authTokenInvalid",
  [AUTH_ACCOUNT_INACTIVE]: "errorCodes.authAccountInactive",
  [AUTH_EMAIL_TAKEN]: "errorCodes.authEmailTaken",
  [AUTH_PASSWORD_EXPIRED]: "errorCodes.authPasswordExpired",
  [AUTH_PASSWORD_MISMATCH]: "errorCodes.authPasswordMismatch",
  [AUTH_PASSWORD_UNAVAILABLE]: "errorCodes.authPasswordUnavailable",
  [AUTHZ_ADMIN_REQUIRED]: "errorCodes.authzAdminRequired",
  [AUTHZ_FORBIDDEN]: "errorCodes.authzForbidden",
  [CREDIT_INSUFFICIENT]: "errorCodes.creditInsufficient",
  [RESOURCE_NOT_FOUND]: "errorCodes.resourceNotFound",
  [RESOURCE_CONFLICT]: "errorCodes.resourceConflict",
  [VALIDATION_ERROR]: "errorCodes.validationError",
  [VALIDATION_LIMIT_EXCEEDED]: "errorCodes.validationLimitExceeded",
  [EXTERNAL_TIMEOUT]: "errorCodes.externalTimeout",
  [EXTERNAL_FAILURE]: "errorCodes.externalFailure",
  [EXTERNAL_PROVIDER_UNAVAILABLE]: "errorCodes.externalProviderUnavailable",
  [INTERNAL_ERROR]: "errorCodes.internalError",
  [NETWORK_ERROR]: "errorCodes.networkError",
  [TIMEOUT]: "errorCodes.timeout",
};

/**
 * Get the i18n key for an error code, or null if none is mapped.
 */
export function getErrorI18nKey(code: string): string | null {
  return ERROR_CODE_I18N_MAP[code] ?? null;
}

/**
 * Get a user-friendly error message for an API error.
 *
 * Strategy:
 * 1. If the error code has a mapped i18n key AND the translation exists, use it.
 * 2. Otherwise fall back to the server-provided message.
 * 3. If neither exists, return the generic "Unknown error" key.
 *
 * @param code    The error code from `errors[].code`
 * @param message The server message from `errors[].message`
 * @param t       The i18next `t` function
 */
export function getErrorMessage(
  code: string,
  message: string,
  t: (key: string, options?: Record<string, unknown>) => string,
): string {
  const i18nKey = getErrorI18nKey(code);
  if (i18nKey) {
    const translated = t(i18nKey, { defaultValue: "" });
    if (translated && translated !== i18nKey) {
      return translated;
    }
  }
  // Fall back to server message
  return message || t("common.unknownError");
}

File: src/api/error_codes.py
"""Standardized API error codes for machine-readable error responses.

Every HTTPException raised by the API should use one of these codes via
the ``api_error()`` helper.  The global exception handler in ``app.py``
extracts the code from the detail dict and returns it in the standard
``ErrorDetail`` envelope.

This module is intentionally flat (no enums, no inheritance) so that
error codes are simple string constants suitable for serialization and
comparison both in Python and in frontend TypeScript.
"""

from __future__ import annotations

from fastapi import HTTPException


class ErrorCode:
    """Machine-readable error code constants."""

    # Authentication
    AUTH_INVALID_CREDENTIALS = "AUTH_INVALID_CREDENTIALS"
    AUTH_TOKEN_EXPIRED = "AUTH_TOKEN_EXPIRED"
    AUTH_TOKEN_INVALID = "AUTH_TOKEN_INVALID"
    AUTH_ACCOUNT_INACTIVE = "AUTH_ACCOUNT_INACTIVE"
    AUTH_EMAIL_TAKEN = "AUTH_EMAIL_TAKEN"
    AUTH_PASSWORD_EXPIRED = "AUTH_PASSWORD_EXPIRED"
    AUTH_PASSWORD_MISMATCH = "AUTH_PASSWORD_MISMATCH"
    AUTH_PASSWORD_UNAVAILABLE = "AUTH_PASSWORD_UNAVAILABLE"
    AUTH_OAUTH_NOT_CONFIGURED = "AUTH_OAUTH_NOT_CONFIGURED"
    AUTH_OAUTH_NOT_IMPLEMENTED = "AUTH_OAUTH_NOT_IMPLEMENTED"
    AUTH_MISSING_AUTH_CODE = "AUTH_MISSING_AUTH_CODE"

    # Authorization
    AUTHZ_ADMIN_REQUIRED = "AUTHZ_ADMIN_REQUIRED"
    AUTHZ_FORBIDDEN = "AUTHZ_FORBIDDEN"
    AUTHZ_API_KEY_INVALID = "AUTHZ_API_KEY_INVALID"

    # Credits
    CREDIT_INSUFFICIENT = "CREDIT_INSUFFICIENT"

    # Resources
    RESOURCE_NOT_FOUND = "RESOURCE_NOT_FOUND"
    RESOURCE_CONFLICT = "RESOURCE_CONFLICT"
    RESOURCE_GONE = "RESOURCE_GONE"

    # Validation
    VALIDATION_ERROR = "VALIDATION_ERROR"
    VALIDATION_EMPTY_UPDATE = "VALIDATION_EMPTY_UPDATE"
    VALIDATION_LIMIT_EXCEEDED = "VALIDATION_LIMIT_EXCEEDED"

    # External services
    EXTERNAL_TIMEOUT = "EXTERNAL_TIMEOUT"
    EXTERNAL_FAILURE = "EXTERNAL_FAILURE"
    EXTERNAL_PROVIDER_UNAVAILABLE = "EXTERNAL_PROVIDER_UNAVAILABLE"

    # Server
    INTERNAL_ERROR = "INTERNAL_ERROR"


def api_error(
    status_code: int,
    code: str,
    message: str,
    field: str | None = None,
) -> HTTPException:
    """Create an HTTPException with a structured error code in the detail.

    The global exception handler in ``app.py`` detects dict-shaped details
    and extracts the ``code`` key, producing a proper ``ErrorDetail``
    envelope.  This helper is syntactic sugar -- routers may also raise
    ``HTTPException(detail={"code": ..., "message": ...})`` directly.
    """
    detail: dict = {"code": code, "message": message}
    if field is not None:
        detail["field"] = field
    return HTTPException(status_code=status_code, detail=detail)
